### PR TITLE
RT4186: Add DSA_dup() function

### PR DESCRIPTION
**Pull request by tmshort**

### CODE REVIEW

File: crypto/dsa/dsa_lib.c [hunk 1]
```diff
@@ -273,3 +273,48 @@ DH *DSA_dup_DH(const DSA *r)
     return NULL;
 }
 #endif
+
+DSA *DSA_dup(const DSA *dsa)
+{
+    DSA *ret = NULL;
+
+    if (dsa == NULL)
```
Comment (richsalz) [inline crypto/dsa/dsa_lib.c]: remove this; crash on NULL.

File: crypto/dsa/dsa_lib.c [hunk 2]
```diff
@@ -273,3 +273,48 @@ DH *DSA_dup_DH(const DSA *r)
...
+        goto err;
+
+    ret = DSA_new_method(dsa->engine);
+    if (ret == NULL)
+        goto err;
+
+    if (dsa->p != NULL)
```
Comment (richsalz) [inline crypto/dsa/dsa_lib.c]: you can merge all each of these into a single if, like
```
  if (dsa->p != NULL && (ret->p=BN_dup()) == NULL)
    goto err;
```

File: crypto/dsa/dsa_lib.c [hunk 3]
```diff
@@ -273,3 +273,48 @@ DH *DSA_dup_DH(const DSA *r)
...
+            goto err;
+    if (dsa->r != NULL)
+        if ((ret->r = BN_dup(dsa->r)) == NULL)
+            goto err;
+
+    if (!CRYPTO_dup_ex_data(CRYPTO_EX_INDEX_DSA, &ret->ex_data, &dsa->ex_data))
+        goto err;
+
+    ret->flags = dsa->flags;
+
```
Comment (richsalz) [inline crypto/dsa/dsa_lib.c]: delete this blank line.

**Comment by richsalz** [comment]
doc update?

File: crypto/dsa/dsa_lib.c (hunk 1 above)
Comment (tmshort) [inline crypto/dsa/dsa_lib.c]: Many other _dup() functions protect against a bad input (EC_POINT_dup, OBJ_dup, etc).

Comment (richsalz) [inline crypto/dsa/dsa_lib.c]: We’re moving away from that.

**Comment by richsalz** [comment]
@tmshort hope you'll be able to address these nits soon; API freeze in a few days...

**Comment by tmshort** [comment]
PR was updated with all but the NULL check removal.

**Comment by tmshort** [comment]
PR just updated with removed NULL check.

**Comment by tmshort** [comment]
Updated CHANGES and libcrypto.num.

**Comment by tmshort** [comment]
In our case, DSAparams_dup() works as desired.

**Pull request closed without merging**